### ISSUE TITLE
ArrayIndexOutOfBoundsException - Upgrade from 0.3.6 to 0.4.2

### ISSUE THREAD

**Issue by kmualem** [zeromq/jeromq#490]
Hi Guys,

As part of the suggestion in #440 in order to solve the OutOfMemory exception, I have upgraded to 0.4.2 and noticed that it breaks the existing functionality that worked in 0.3.6 :(

The #440 is very critical as we are using it in our production system and it crashed from time to time.

This is the stack trace:

```
java.lang.ArrayIndexOutOfBoundsException: 7
        at zmq.socket.pubsub.Mtrie.rmHelper(Mtrie.java:254)
        at zmq.socket.pubsub.Mtrie.rm(Mtrie.java:238)
        at zmq.socket.pubsub.XPub.xreadActivated(XPub.java:104)
        at zmq.SocketBase.readActivated(SocketBase.java:1099)
        at zmq.pipe.Pipe.processActivateRead(Pipe.java:312)
        at zmq.ZObject.processCommand(ZObject.java:51)
        at zmq.Command.process(Command.java:75)
        at zmq.SocketBase.processCommands(SocketBase.java:937)
        at zmq.SocketBase.getSocketOpt(SocketBase.java:251)
        at zmq.poll.PollItem.readyOps(PollItem.java:115)
        at zmq.ZMQ.poll(ZMQ.java:709)
        at org.zeromq.ZMQ$Poller.poll(ZMQ.java:2939)
        at org.zeromq.ZLoop.start(ZLoop.java:263)
```

It seems the same issue as in ticket #476  

Can you please approve that this issue is fixed ? 
Can you please let me what is the estimation for 0.4.3 ? As it seems like a very basic functionality is now broken

Thanks for you help!

_(EDIT: stacktrace formatting - @daveyarwood)_

**Comment by fredoboulo** [comment]
Hello,

It is the same stack trace, so it should be related to #476, however it does not seem coming from the same usage. Could you provide some minimal test case that can repeat the error?
If you could also try with the latest master code, that could help to assess the situation, as #476 is already considered as fixed on master branch.

@trevorbernard, is there any agenda for a new release? A lot of fixes have been performed since 0.4.2.

**Comment by kmualem** [comment]
Thanks @fredoboulo 

I'm not sure yet how to reproduce this error. Our QA found it during regression tests on the product, the only change that I did from 0.3.6 to 0.4.2 is to add the context to the Zloop constructor.

The reactor thread in our code handle the following:
1. Incoming events in Router socket that arrive from Dealer sockets. Then it sends the information downstream to the XPUB socket.
2. Incoming events in XPUB socket that arrive from XSUB sockets. Then sometimes it sends the information to a PUSH socket with multiple PULL workers.

From the stack trace it seems like it's related to the XPUB socket. Can you please explain in what scenarios it should arrive to this code ? It may help me to think on a test case... and of course, I'll try to check it with the latest master code!

 It will be very helpful to get some details about the next release...

Thanks for your help!

**Comment by fredoboulo** [comment]
It would depend on the dynamic of your subscriptions/unsubscriptions, basically the bug you encountered occurs when unsubscription arrives to the XPUB socket, which happens when (X)SUB socket is closed or unsubscribe(topic) is called.
Tests that assess the fix are src/test/java/org/zeromq/XpubXsubZTest.java and  src/test/java/zmq/socket/pubsub/XpubXsubTest.java in the master code.

For next release, I'm not in charge of that, @trevorbernard may have a better visibility on the agenda or could redirect to someone who is, but it would basically occur when there are enough fixes to justify a new release and when the code is considered as stable enough. Don't be shy to consider yourself and your organization part of the latter (or the former) :)

**Comment by kmualem** [comment]
I checked the master code and it seems like the problem wasn't reproduced in our regression tests.

Do we have an estimation for the 1.0.43 release?

**Comment by fredoboulo** [comment]
Estimation is ... now :)

**Comment by kmualem** [comment]
Hi,

I have checked our applications using the new 0.43 release and I'm getting a failure upon the process shutdown.

The only change that I did while moving from 0.3.6 to 0.4.3 is to add the context to the reactor ZLoop constructor.

The order that I'm destroying the resources is:
1. Remove all pollers from the reactor (ZLoop)
2. destroy the reactor (org.zeromq.ZLoop.destroy())
3. destory the sockets by zContext.destroySocket(socket)
4. destroy the context.

I noticed that in 0.4.3 the ZLoop destroy already close the zcontext. So I tried also to stop first the sockets and then the reactor... but it didn't help.

Can you please help?

I really need to upgrade to this version in order to solve "port scanning" issue - #440

**Comment by fredoboulo** [comment]
That may be a mistake to close the context here. destroy() was previously doing nothing, so if we revert back to that state, there is actually no need for that method.
Can you try to omit the step 2. ?

I can provide no explanation or motivation for closing the context there, apart from the excuse that it was part of a huge refactoring. Please accept my apologies on that one.

**Comment by kmualem** [comment]
@fredoboulo  - Thanks for the quick response!

I can try to omit step 2, but I'm trying to understand how this issue was not discovered before :( and if this is the only problem while using the same context in multiple ZLoop instances... 
In case of multiple threads that use ZLoop, the first thread that closes the reactor (ZLoop) will close the underlying context and it's wrong of course.

In addition, I'm trying to understand what is the usage of the Context in the ZLoop?
I see this row:
pollset = context.poller(pollSize);

Is it safe to have multiple threads, each one with its ZLoop and inside the ZLoop, pollers use the same context? 

Any idea who can help me with this issue?

Thanks for your help.

**Comment by fredoboulo** [comment]
About the usage of context in ZLoop:

Context in constructor was added one year ago, to provide a way to create a Poller in the ZLoop.
The whole point is that polling needs a Java Selector, which needs to be closed afterwards. Context ensures that by keeping references to created selectors, so there is no memory leak when releasing resources.

@daveyarwood may be of some help as well.

As far as I can see, there are only 2 tests for ZLoop (I would even say only one), coverage may be improved with your help!

**Comment by kmualem** [comment]
I opened a new issue for this problem #505 

I will test again the original issue of this ticket (ArrayIndexOutOfBoundsException ) and will update if it works now

**Comment by kmualem** [comment]
We checked the ArrayIndexOutOfBoundsException and it seems that it was fixed in 0.4.3, So I'm closing this ticket.
